### ISSUE TITLE
TypeScript: TFunction returning "object" rather than the default "string" type

### ISSUE THREAD

**Issue by totszwai** [i18next/react-i18next#1466]
I have the following property defined in my interface:
```
export interface Blah {
  primary?: string | number;
}
```

When passing the translated caption over to the property, it would throw TypeScript error.

<img width="781" alt="image" src="https://user-images.githubusercontent.com/4440365/157533191-f63ad556-c742-4e99-8093-a38303507af6.png">

Is there a way or workaround for this? I don't think there is a way to even type it to return string either? The best we could configure it is just the namespace and option... there isn't a way to deep pass it to force the `TDefaultResult`?

```
const { t } = useTranslation<'BlahBlah', SomeOption>();
```

**Comment by pedrodurek** [comment]
Hey @totszwai, I didn't understand what you're trying accomplish, could you provide more info or if possible an example (on codesanbox)?

**Comment by totszwai** [comment]
@pedrodurek There are 2 definitions for the `t` function, one returns a `string` which is perfectly fine, but the other returns an `object`. In the screenshot below, the +1 overload is the one returning an `object`.

![image](https://user-images.githubusercontent.com/4440365/157580846-39fdb953-8aba-4cc9-96a3-2e77a61fef89.png)

In TypeScript, if the receiving end of the property is typed to `string`, everything is fine, it would not complaint.

However, if the property is anything other than `string` in my case, it was `string | number`, then i18next-react (or i18next)'s definition would ended up returning the definition of `object` (instead of `string`) and that's what you would see in the TypeScript error posted in the original post...

It would say `object` is not compatible with `string | number` but had it returned with the `string` signature, everything would worked out.

There is no way for me to tell the `t` function to always return as `string`

**Comment by pedrodurek** [comment]
Hey @totszwai, `t` function can return object if you have `returnObjects` config set to `true`. So typescript will automatically infer the `t` function return type according to the `key` passed. For example:
```json
{
  bla: {
    blabla: 'bla'
  }
}
```

```ts
const {t} = useTranslation();

t('bla') // will infer as object
t('bla.blabla') // will infer as string
```

I still don't get it what are you trying to accomplish, why do you need to force both generics to be `BlahBlah` as `SomeOption` (`const { t } = useTranslation<'BlahBlah', SomeOption>();`), typescript will automatically infer according to the namespace and options passed.

**Comment by totszwai** [comment]
@pedrodurek So what did I do wrong?

This is my locale JSON file, `network.json`, I have a namespace "network" setup for it, the file itself has a flat structure.\:
```
{
  "alert": "Alert",
  "alerts": "Alerts",
  ...
}
```

So using it as described, why is it returning an object?
```
const {t} = useTranslation('network');

t('alert');
```

Like I said:
```
const test1: string;
const test2: string | number;

test1 = t('alert'); // OK
test2 = t('alert'); // FAILED, resulting that tslint error because t was returning and object.
```

**Comment by pedrodurek** [comment]
maybe something wrong with your `tsconfig` or `tslint`, here it's working fine. https://codesandbox.io/s/react-i18next-typed-keys-bug-forked-x77zq?file=/src/app.tsx

**Comment by totszwai** [comment]
Despite using the exact same version of packages in your sandbox, the TypeScript hinting returned by your sandbox and inside VScode is different...

In the sandbox, is returning:
```
<key, string, StringMap> // the important one is the 2nd parameter
```
<img width="593" alt="image" src="https://user-images.githubusercontent.com/4440365/159301471-52076b33-9ef0-4b63-8dc5-cbb7d907d490.png">

However, if we observe inside VScode environment, is `TFunctionResult` as the second parameter...
```
<key, TFunctionResult, StringMap> 
```
<img width="709" alt="image" src="https://user-images.githubusercontent.com/4440365/159301847-e2205d5c-2891-4b1a-9768-492178245d44.png">


Has anyone seen such an issue before? And go away stalebot! :P

**Comment by akunzai** [comment]
I meet the same TS error after I bump @types/react version from 17 to 18

If I revert the @types/react version to 17.0.44 and refresh, the TS errors gone.

I'm not sure what makes the TS error.

So I open a [discussion](https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/59759).

and create a [mini reproduce](https://stackblitz.com/edit/react-ts-otzsat?file=index.tsx).

**Comment by akunzai** [comment]
Workaround: Force cast the `t` function result to the string type

```jsx
<h1>{t('Hello') as string}</h1>
```

**Comment by totszwai** [comment]
I want to thank @akunzai for creating the mini reproduction! Yeah, I know there is a workaround (is all just TypeScript and JS at the end of the day, heck we could cast everything to `any`), but the whole point is not to have the workaround. Is incredibly annoying and so much overhead just to cast every single translation you have you do.

**Comment by akunzai** [comment]
I will delete all my previous comments on this issue because my issue was resolved by #1480,
But I'm not quite sure #1480 can resolve this issue too.

**Comment by totszwai** [comment]
@akunzai Is there a way to try that custom build that you just merged?

**Comment by akunzai** [comment]
It included since v11.16.4 release, no custom build needed.
